_posts/contests/2021-02-05-Performance-analysis-and-tuning-contest-5.md

```diff
@@ -0,0 +1,216 @@
+
+The fifth edition of the contest will be run by Ivica Bogosavljevic from [Johny's Software Lab](https://johnysswlab.com) blog. Ivica also writes about software performance, so feel free to go and check out his blog, there is a ton of useful content there.
+
+The benchmark for the 5th edition is KALDI. Kaldi is an open-source toolkit for speech recognition written in C++ intended to be used by researchers.
+
+In the essence, Kaldi takes an input model and recorded speech and then it converts it to a textual representation. Speech recognition is a complex topic, but as always, we focus on hardware efficiency and general data processing efficiency, not on the speech recognition algorithm itself.
+
+## Quickstart
+
+Here are instructions on how to build Kaldi on Linux. To download and build kaldi for the first time do the following:
+
+```bash
+$ git clone https://github.com/kaldi-asr/kaldi.git
+$ cd kaldi
+$ git checkout ca6d133262aa183b23f6daba48995bd7576fb572
+$ cd tools
+$ extras/check_dependencies.sh
+$ make -j8
+$ extras/install_openblas.sh
+$ cd OpenBLAS
+$ make PREFIX=`pwd` install
+$ cd ../../src/
+$ ./configure --shared --openblas-root=../tools/OpenBLAS/
+```
+The configure script generates the file called `kaldi.mk`. Open the file, find the line that starts with `CXXFLAGS`, and replace default optimization level `-O1` with `-O3`,
+
+```bash
+$ make clean -j8
+$ make depend -j8
+$ make -j8
+```
+
+The first time you build kaldi it can take a lot of time to compile everything, so it is a good idea to let it compile and go do something else. Later, when doing incremental builds, just run `make -j8`. If you added or removed headers (changed dependencies), you should also run `make depend -j8` before `make -j8`.
+
+## Downloading and running the benchmark
+
+Under the assumption you are in `kaldi/src` directory, to download the test you will do the following;
+
+```bash
+$ wget https://johnysswlab.com/downloads/test-speed.zip
+$ unzip test-speed.zip
+```
+
+To run the benchmark, execute:
+
+```bash
+$ cd test-speed
+../online2bin/online2-wav-nnet3-latgen-faster --word-symbol-table=graph/words.txt --config=conf/model.conf am/final.mdl graph/HCLG.fst ark:test.utt2spk scp:test.scp ark:/dev/null
+```
+
+To verify the result of the change:
+
+```bash
+TODO
+```
+
+## Rules of the game
+
+When you profile the example, you will notice that a lof time is spent in `sgemm_kernel` and `sgemm_copy` functions. Unfortunately, these functions do not belong to kaldi, they belong to OpenBLAS library and they are not the goal of optimization in this contest.
+
+To limit the scope of the changes, you are allowed to:
+* Modify only files: `lattice-faster-decoder.cc` and `lattice-faster-decoder.h`, files which are part of kaldi
```

Comment:
Giving too many hints, no?
I tend to prefer more generic statements like "You cannot change the core of the algorithm". Sometimes folks can come up with some really clever ideas, which they couldn't implement due to our limitations.

Reply:
I wrote down that we can change the rules if needed. `Please note that this rule is not written in stone, if you believe that for good performance it is necessary to change other files as well, let us know and we can agree to change this rule.`